File: 0498-diagonal-traverse/0498-diagonal-traverse.py
class Solution(object):
    def findDiagonalOrder(self, mat):
        flag=1
        n=len(mat)
        m=len(mat[0])
        length=len(mat)*len(mat[0])
        result = []
        row,col=0,0
        for i in range(0,length):
            result.append(mat[row][col])
            if(flag):
                if col==m-1:
                    row+=1
                    flag=0

                elif row==0:
                    col+=1
                    flag=0 
                else:
                    row-=1
                    col+=1

            else:
                if row==n-1:
                    col+=1
                    flag=1
                elif col==0:
                    row+=1
                    flag=1
                else:
                    row+=1
                    col-=1     
        return result
                

File: 0498-diagonal-traverse/0498-diagonal-traverse.java
class Solution {
    public int[] findDiagonalOrder(int[][] mat) {
        int n = mat.length;
        int m = mat[0].length;
        int[] result = new int[n * m];

        int row = 0, col = 0, idx = 0;
        boolean flag = true;

        while (idx < n * m) {
            result[idx++] = mat[row][col];

            if (flag) 
            {
                if (col == m - 1) 
                {
                    row++;
                    flag = false;
                } 
                else if (row == 0) 
                {
                    col++;
                    flag = false;
                } 
                else
                {
                    row--;
                    col++;
                }
            } 
            else 
            { 
                if (row == n - 1) 
                {
                    col++;
                    flag = true;
                } 
                else if (col == 0)
                {
                    row++;
                    flag = true;
                } else 
                {
                    row++;
                    col--;
                }
            }
        }

        return result;
    }
}
